File: apps/apple/Sources/SharedAppleRemoteCore/LinkClient.swift
import CryptoKit
import Foundation

/// Every way a link call can fail, mirroring the TypeScript reference
/// client's `LinkError` vocabulary.
public enum LinkClientError: Error, Equatable {
    /// The carrier failed before an HTTP response or returned a non-2xx status
    /// that is not an authorization refusal.
    case carrier(status: Int, message: String)
    /// A paired identity is required but none is persisted.
    case unpaired
    /// Carrier authorization or a Remote operation refused the request.
    case refused(code: String, message: String)
    /// The envelope or frame bytes were not decodable.
    case badWire(String)
}

/// The link-client state machine: pair once, then describe, call unary
/// endpoints through the shared `/api` chain, and open NDJSON Remote streams
/// — the same sequences the TypeScript reference client (`dsh-link-client`)
/// exercises for the pair → connect → session → prompt → approval →
/// reconnect vertical slice.
public final class LinkClient {
    private let baseURL: URL
    private let session: URLSession
    private let store: LinkCredentialsStoring
    private let pinned: String

    /// - Parameters:
    ///   - baseURL: the carrier endpoint from the pairing payload.
    ///   - pinnedFingerprint: SPKI fingerprint the TLS handshake pins.
    ///   - store: where the paired identity persists.
    /// The SPKI fingerprint this client pins; tests and restore read it.
    public var pinnedFingerprint: String { pinned }

    public init(baseURL: URL, pinnedFingerprint: String, store: LinkCredentialsStoring) {
        self.baseURL = baseURL
        self.store = store
        self.pinned = pinnedFingerprint
        self.session = URLSession(
            configuration: .ephemeral,
            delegate: LinkPinningDelegate(pinnedFingerprint: pinnedFingerprint),
            delegateQueue: nil
        )
    }

    /// The persisted identity, or nil before the first successful pairing.
    public var credentials: LinkCredentials? { store.load() }

    /// Pair with a host by exchanging the one-time QR code for a durable
    /// identity. Generates a fresh Ed25519 key whose SPKI DER the host
    /// stores; persists the returned identity.
    /// - Parameters:
    ///   - payload: the scanned pairing payload.
    ///   - deviceName: the user-chosen name shown in the host's device list.
    /// - Returns: the persisted credentials.
    public func pair(payload: LinkPairingPayload, deviceName: String) async throws -> LinkCredentials {
        guard Self.pairingPayloadOwnsTransport(payload, baseURL: baseURL, pinnedFingerprint: pinned) else {
            throw LinkClientError.badWire("pairing payload does not own this client transport")
        }
        let key = Curve25519.Signing.PrivateKey()
        let request = LinkPairRequest(
            code: payload.code,
            deviceName: deviceName,
            devicePublicKey: LinkSigning.ed25519SpkiDer(publicKeyRaw: key.publicKey.rawRepresentation).base64EncodedString()
        )
        let body = try JSONEncoder().encode(request)
        let data = try await post(path: "/link/pair", body: body, signed: false)
        let value = try Self.decode(LinkPairResponse.self, from: data)
        let credentials = LinkCredentials(
            deviceId: value.deviceId,
            hostId: value.hostId,
            hostName: value.hostName,
            role: value.role.rawValue,
            endpoint: payload.endpoint,
            pinnedFingerprint: payload.spkiFingerprint,
            signingKeyBase64: key.rawRepresentation.base64EncodedString()
        )
        store.save(credentials)
        return credentials
    }

    /// Ask the authenticated host for its description and capabilities.
    public func describe() async throws -> LinkHostDescription {
        let data = try await post(path: "/link/describe", body: Data(), signed: true)
        return try Self.decode(LinkHostDescription.self, from: data)
    }

    /// Call one unary Remote endpoint through the shared `/api` chain.
    /// - Parameters:
    ///   - method: canonical endpoint, for example `session/list`.
    ///   - args: named wire arguments.
    /// - Returns: the business value on success.
    /// - Throws: `LinkClientError.unpaired` or `badWire` for unusable local
    ///   credentials or response bytes, `carrier` for transport failures, and
    ///   `refused` for carrier authorization or business-call refusal.
    public func call(
        _ method: String,
        args: [String: LinkJsonValue] = [:]
    ) async throws -> LinkJsonValue {
        let rpcId = "rpc-\(UUID().uuidString)"
        let envelope = LinkRpcRequestEnvelope(
            type: "client-request",
            rpcId: rpcId,
            method: method,
            payload: LinkRpcPayload(args: args)
        )
        let body = try JSONEncoder().encode(envelope)
        let data = try await post(path: "/api/\(method)", body: body, signed: true)
        let response = try Self.decode(LinkRpcResponseEnvelope.self, from: data)
        return try Self.value(from: response, expectedRpcId: rpcId)
    }

    /// Open one NDJSON Remote stream. A carrier authorization refusal throws
    /// before the stream is returned; a typed failure frame finishes the
    /// returned stream with `refused`. A transport failure mid-stream surfaces
    /// as the underlying `URLError` so callers resubscribe rather than treat
    /// silence as completion.
    /// - Parameters:
    ///   - endpoint: canonical stream endpoint, for example `$events`.
    ///   - payload: the stream's opening payload arguments.
    /// - Returns: an async stream of decoded frame values.
    /// - Throws: `LinkClientError.unpaired` or `badWire` for unusable local
    ///   credentials or response bytes, `carrier` for transport failures, and
    ///   `refused` when carrier authorization rejects the stream request.
    public func stream(
        _ endpoint: String,
        payload: [String: LinkJsonValue] = [:]
    ) async throws -> AsyncThrowingStream<LinkJsonValue, Error> {
        let body = try JSONEncoder().encode(LinkStreamRequest(args: payload))
        let data: (bytes: URLSession.AsyncBytes, response: URLResponse)
        do {
            var request = try URLRequest(url: Self.url(base: baseURL, path: "/link/stream/\(endpoint)"))
            request.httpMethod = "POST"
            request.setValue("application/json", forHTTPHeaderField: "content-type")
            try Self.applyCredentials(to: &request, body: body, store: store)
            request.httpBody = body
            data = try await session.bytes(for: request)
        } catch let error as LinkClientError {
            throw error
        } catch {
            throw LinkClientError.carrier(status: 0, message: String(describing: error))
        }
        try await Self.checkStreamResponse(response: data.response, bytes: data.bytes)
        return AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    for try await line in data.bytes.lines {
                        guard let frameData = line.data(using: .utf8), !frameData.isEmpty else { continue }
                        let frame = try Self.decode(LinkStreamFrame.self, from: frameData)
                        if frame.k == .e {
                            continuation.finish(throwing: LinkClientError.refused(
                                code: frame.c ?? "internal",
                                message: frame.m ?? "stream failed"
                            ))
                            return
                        }
                        continuation.yield(frame.v ?? .null)
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    /// Forget the paired identity; the host refuses the next request.
    public func unpair() {
        store.clear()
    }

    /// Rebuild the paired client from persisted credentials — the relaunch
    /// path that skips pairing and pins the stored fingerprint again.
    /// - Parameter store: where the identity persists.
    /// - Returns: the client, or nil before the first successful pairing or
    ///   when the stored endpoint no longer parses.
    public static func restore(store: LinkCredentialsStoring) -> LinkClient? {
        guard let credentials = store.load() else { return nil }
        guard let endpoint = URL(string: credentials.endpoint) else { return nil }
        return LinkClient(
            baseURL: endpoint,
            pinnedFingerprint: credentials.pinnedFingerprint,
            store: store
        )
    }

    // MARK: - Internals

    /// Join the carrier base URL with an absolute request path.
    static func url(base: URL, path: String) -> URL {
        let trimmed = base.absoluteString.hasSuffix("/")
            ? String(base.absoluteString.dropLast())
            : base.absoluteString
        return URL(string: trimmed + path)!
    }

    /// Whether one pairing payload names the endpoint and pin this client uses.
    static func pairingPayloadOwnsTransport(
        _ payload: LinkPairingPayload,
        baseURL: URL,
        pinnedFingerprint: String
    ) -> Bool {
        guard let payloadURL = URL(string: payload.endpoint) else { return false }
        let normalizedPayload = payloadURL.absoluteString.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let normalizedBase = baseURL.absoluteString.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        return normalizedPayload == normalizedBase && payload.spkiFingerprint == pinnedFingerprint
    }

    /// Send one unary request, optionally device-signed, checking the status.
    private func post(path: String, body: Data, signed: Bool) async throws -> Data {
        do {
            var request = try URLRequest(url: Self.url(base: baseURL, path: path))
            request.httpMethod = "POST"
            request.setValue("application/json", forHTTPHeaderField: "content-type")
            if signed {
                try Self.applyCredentials(to: &request, body: body, store: store)
            }
            request.httpBody = body
            let (data, response) = try await session.data(for: request)
            try Self.check(response: response, data: data)
            return data
        } catch let error as LinkClientError {
            throw error
        } catch {
            throw LinkClientError.carrier(status: 0, message: String(describing: error))
        }
    }

    /// Attach the three credential headers with a fresh timestamp signature.
    static func applyCredentials(to request: inout URLRequest, body: Data, store: LinkCredentialsStoring) throws {
        guard let credentials = store.load() else { throw LinkClientError.unpaired }
        guard let privateKey = Data(base64Encoded: credentials.signingKeyBase64) else {
            throw LinkClientError.badWire("stored signing key is not base64")
        }
        let timestamp = String(Int(Date().timeIntervalSince1970 * 1000))
        let path = request.url?.path ?? "/"
        let input = LinkSigning.signingInput(
            timestamp: timestamp,
            method: request.httpMethod ?? "POST",
            path: path,
            bodySha256Hex: LinkSigning.sha256Hex(body)
        )
        request.setValue(credentials.deviceId, forHTTPHeaderField: LinkSigning.deviceIdHeader)
        request.setValue(timestamp, forHTTPHeaderField: LinkSigning.timestampHeader)
        request.setValue(try LinkSigning.sign(input: input, privateKeyRaw: privateKey), forHTTPHeaderField: LinkSigning.signatureHeader)
    }

    /// Classify a non-2xx carrier answer. Only an HTTP 403 with the JSON string
    /// `error` equal to `forbidden` is an authorization refusal; all other
    /// statuses remain carrier failures.
    static func check(response: URLResponse, data: Data?) throws {
        guard let http = response as? HTTPURLResponse else { return }
        guard (200..<300).contains(http.statusCode) else {
            let document: [String: Any]? = data.flatMap { body in
                guard let value = try? JSONSerialization.jsonObject(with: body) else { return nil }
                return value as? [String: Any]
            }
            let message = nonemptyString("message", in: document)
                ?? nonemptyString("reason", in: document)
                ?? "HTTP \(http.statusCode)"
            if http.statusCode == 403, document?["error"] as? String == "forbidden" {
                throw LinkClientError.refused(code: "forbidden", message: message)
            }
            throw LinkClientError.carrier(status: http.statusCode, message: message)
        }
    }

    /// Read and classify an unsuccessful stream response without consuming a
    /// successful stream's first byte.
    static func checkStreamResponse<Bytes: AsyncSequence>(
        response: URLResponse,
        bytes: Bytes
    ) async throws where Bytes.Element == UInt8 {
        guard let http = response as? HTTPURLResponse,
              !(200..<300).contains(http.statusCode) else { return }
        var body = Data()
        do {
            for try await byte in bytes {
                body.append(byte)
            }
        } catch {
            throw LinkClientError.carrier(status: http.statusCode, message: "HTTP \(http.statusCode)")
        }
        try check(response: http, data: body)
    }

    private static func nonemptyString(_ field: String, in document: [String: Any]?) -> String? {
        guard let value = document?[field] as? String, !value.isEmpty else { return nil }
        return value
    }

    /// Decode or map to `badWire`.
    static func decode<T: Decodable>(_ type: T.Type, from data: Data) throws -> T {
        do {
            return try JSONDecoder().decode(T.self, from: data)
        } catch {
            throw LinkClientError.badWire(String(describing: error))
        }
    }

    /// Validate one generated unary response and project its business value.
    /// - Parameters:
    ///   - response: decoded canonical response envelope.
    ///   - expectedRpcId: request identity the Host must echo.
    /// - Returns: the returned JSON value, or `.null` for a successful void RPC.
    static func value(from response: LinkRpcResponseEnvelope, expectedRpcId: String) throws -> LinkJsonValue {
        guard response.type == "server-response" else {
            throw LinkClientError.badWire("unexpected response type \(response.type)")
        }
        guard response.rpcId == expectedRpcId else {
            throw LinkClientError.badWire("rpcId mismatch")
        }
        if response.result.ok {
            guard response.result.error == nil else {
                throw LinkClientError.badWire("successful result carried an error")
            }
            return response.result.value ?? .null
        }
        guard response.result.value == nil, let failure = response.result.error else {
            throw LinkClientError.badWire("failed result lacked a structured error")
        }
        throw LinkClientError.refused(code: failure.code, message: failure.message)
    }
}

File: apps/android/core/src/main/kotlin/ai/deepseek/dsh/link/LinkClient.kt
package ai.deepseek.dsh.link

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.async
import kotlinx.coroutines.cancelAndJoin
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.supervisorScope
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import okhttp3.Call
import okhttp3.Callback
import okhttp3.EventListener
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import java.io.IOException
import java.security.KeyPairGenerator
import java.util.Base64
import java.util.UUID
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

/** Every way a link call can fail, mirroring the Swift `LinkClientError`. */
sealed class LinkClientException(message: String) : RuntimeException(message) {
    class Carrier(val status: Int, message: String) : LinkClientException("carrier $status: $message")

    class Unpaired : LinkClientException("no paired identity")

    class Refused(val code: String, message: String) : LinkClientException("refused $code: $message")

    class BadWire(message: String) : LinkClientException("bad wire: $message")
}

/**
 * Deployment-owned HTTP timeouts for the Link transport. Zero disables the
 * corresponding OkHttp timeout, which is required for idle long-lived streams.
 * @param connectTimeoutMillis TCP/TLS connection timeout.
 * @param writeTimeoutMillis request-body write timeout.
 * @param unaryReadTimeoutMillis response-read timeout for pair and unary calls.
 * @param unaryCallTimeoutMillis whole-call timeout for pair and unary calls.
 * @param streamReadTimeoutMillis idle response-read timeout for streams.
 * @param streamCallTimeoutMillis whole-call timeout for streams.
 */
data class LinkTransportConfig(
    val connectTimeoutMillis: Long,
    val writeTimeoutMillis: Long,
    val unaryReadTimeoutMillis: Long,
    val unaryCallTimeoutMillis: Long,
    val streamReadTimeoutMillis: Long,
    val streamCallTimeoutMillis: Long,
) {
    init {
        val values = listOf(
            connectTimeoutMillis,
            writeTimeoutMillis,
            unaryReadTimeoutMillis,
            unaryCallTimeoutMillis,
            streamReadTimeoutMillis,
            streamCallTimeoutMillis,
        )
        require(values.all { it >= 0 }) { "Link transport timeouts must be non-negative" }
    }
}

/** Transport-phase observation used by owner-level cancellation checks. */
internal interface LinkCallObserver {
    fun callStart(path: String) = Unit

    fun requestBodyStart(path: String)

    fun requestBodyEnd(path: String, byteCount: Long)

    fun callFailed(path: String, failure: IOException) = Unit
}

/**
 * The link-client state machine over one owned OkHttp transport — the Kotlin
 * mirror of the Swift `LinkClient`: pair once, then describe, call unary
 * endpoints through the shared `/api` chain, and open NDJSON Remote streams.
 * Every HTTPS call installs [LinkPinning] before sending its request body;
 * [close] cancels all created calls and retires the transport resources.
 */
class LinkClient private constructor(
    baseUrl: String,
    pinnedFingerprint: String,
    private val store: LinkCredentialsStoring,
    transportConfig: LinkTransportConfig,
    callObserver: LinkCallObserver?,
) : java.io.Closeable {
    constructor(
        baseUrl: String,
        pinnedFingerprint: String,
        store: LinkCredentialsStoring,
        transportConfig: LinkTransportConfig,
    ) : this(baseUrl, pinnedFingerprint, store, transportConfig, null)

    private val base: String = baseUrl.trimEnd('/')
    private val pinned: String = pinnedFingerprint
    private val trustManager = LinkPinning.trustManager(pinnedFingerprint)
    private val sslContext = LinkPinning.sslContext(trustManager)
    private val hostnameVerifier = LinkPinning.hostnameVerifier(pinnedFingerprint)
    private val transportClient = OkHttpClient.Builder()
        .sslSocketFactory(sslContext.socketFactory, trustManager)
        .hostnameVerifier(hostnameVerifier)
        .connectTimeout(transportConfig.connectTimeoutMillis, TimeUnit.MILLISECONDS)
        .writeTimeout(transportConfig.writeTimeoutMillis, TimeUnit.MILLISECONDS)
        .apply {
            if (callObserver != null) {
                eventListener(
                    object : EventListener() {
                        override fun callStart(call: Call) {
                            callObserver.callStart(call.request().url.encodedPath)
                        }

                        override fun requestBodyStart(call: Call) {
                            callObserver.requestBodyStart(call.request().url.encodedPath)
                        }

                        override fun requestBodyEnd(call: Call, byteCount: Long) {
                            callObserver.requestBodyEnd(call.request().url.encodedPath, byteCount)
                        }

                        override fun callFailed(call: Call, ioe: IOException) {
                            callObserver.callFailed(call.request().url.encodedPath, ioe)
                        }
                    },
                )
            }
        }
        .build()
    private val unaryClient = transportClient.newBuilder()
        .readTimeout(transportConfig.unaryReadTimeoutMillis, TimeUnit.MILLISECONDS)
        .callTimeout(transportConfig.unaryCallTimeoutMillis, TimeUnit.MILLISECONDS)
        .build()
    private val streamClient = transportClient.newBuilder()
        .readTimeout(transportConfig.streamReadTimeoutMillis, TimeUnit.MILLISECONDS)
        .callTimeout(transportConfig.streamCallTimeoutMillis, TimeUnit.MILLISECONDS)
        .build()
    private val lifecycleLock = Any()
    private val activeCalls = mutableSetOf<TrackedCall>()
    private var closed = false
    private val transportRetired = AtomicBoolean(false)

    private class TrackedCall(
        val call: Call,
        private val retireCall: (Call) -> Unit,
    ) {
        val settled = CompletableDeferred<Unit>()
        private val retiring = AtomicBoolean(false)

        fun retire() {
            if (retiring.compareAndSet(false, true)) retireCall(call)
        }
    }

    private class LinkRetiredCancellation : CancellationException("Link client is closed")

    /** The persisted identity, or null before the first successful pairing. */
    val credentials: LinkCredentials? get() = store.load()

    /** The fingerprint this client was constructed to pin. */
    val pinnedFingerprint: String get() = pinned

    /**
     * Pair with a host by exchanging the one-time QR code for a durable
     * identity. The client generates a fresh Ed25519 key whose SPKI DER the
     * host stores, then validates required identity, role, and protocol fields
     * before the store persists credentials. Work runs off the caller's
     * dispatcher, and cancellation cancels the owned OkHttp call.
     */
    suspend fun pair(payload: LinkPairingPayload, deviceName: String): LinkCredentials = withContext(Dispatchers.IO) {
        if (payload.endpoint.trimEnd('/') != base || payload.spkiFingerprint != pinned) {
            throw LinkClientException.BadWire("pairing payload does not own this client transport")
        }
        val key = KeyPairGenerator.getInstance("Ed25519").generateKeyPair()
        val publicRaw = key.public.encoded.copyOfRange(key.public.encoded.size - 32, key.public.encoded.size)
        val body = Json.encodeToString(
            JsonElement.serializer(),
            kotlinx.serialization.json.buildJsonObject {
                put("code", payload.code)
                put("deviceName", deviceName)
                put("devicePublicKey", Base64.getEncoder().encodeToString(LinkSigning.ed25519SpkiDer(publicRaw)))
            },
        )
        val data = post("/link/pair", body.toByteArray(Charsets.UTF_8), signed = false)
        val document = runCatching { Json.parseToJsonElement(data.decodeToString()).jsonObject }
            .getOrElse { throw LinkClientException.BadWire("pair response is not a JSON object") }
        val value = document.pairResponse(expectedProtocolVersion = payload.v)
        val credentials = LinkCredentials(
            deviceId = value.deviceId,
            hostId = value.hostId,
            hostName = value.hostName,
            role = value.role.wire,
            endpoint = payload.endpoint,
            pinnedFingerprint = payload.spkiFingerprint,
            signingKeyBase64 = Base64.getEncoder().encodeToString(
                key.private.encoded.copyOfRange(key.private.encoded.size - 32, key.private.encoded.size),
            ),
        )
        store.save(credentials)
        credentials
    }

    /** Ask the authenticated host for its description without blocking the caller's dispatcher.
     * Cancellation cancels the owned OkHttp call. */
    suspend fun describe(): LinkHostDescription = withContext(Dispatchers.IO) {
        val data = post("/link/describe", ByteArray(0), signed = true)
        mapHostDescription(Json.parseToJsonElement(data.decodeToString()).jsonObject)
    }

    /**
     * Call one unary Remote endpoint through the shared `/api` chain;
     * throws [LinkClientException.Refused] when carrier authorization or the
     * business call refuses the operation.
     * Coroutine cancellation cancels the owned OkHttp call.
     */
    suspend fun call(method: String, args: Map<String, WireValue> = emptyMap()): WireValue =
        withContext(Dispatchers.IO) {
            val rpcId = "rpc-${UUID.randomUUID()}"
            val envelope = LinkRequestEnvelope(rpcId = rpcId, method = method, args = args)
            val body = Json.encodeToString(envelope.toJsonElement())
                .toByteArray(Charsets.UTF_8)
            val data = post("/api/$method", body, signed = true)
            val response = LinkResponseEnvelope.fromJsonElement(Json.parseToJsonElement(data.decodeToString()))
            if (response.type != "server-response") {
                throw LinkClientException.BadWire("unexpected response type ${response.type}")
            }
            if (response.rpcId != rpcId) throw LinkClientException.BadWire("rpcId mismatch")
            if (response.result.ok) {
                if (response.result.errorCode != null) {
                    throw LinkClientException.BadWire("successful result carried an error")
                }
                return@withContext response.result.value ?: WireValue.NullValue
            }
            if (response.result.value != null) {
                throw LinkClientException.BadWire("failed result carried a value")
            }
            if (response.result.errorCode != null) {
                throw LinkClientException.Refused(response.result.errorCode!!, response.result.errorMessage ?: "")
            }
            throw LinkClientException.BadWire("failed result lacked a structured error")
        }

    /**
     * Open one NDJSON Remote stream: value frames flow as they arrive; carrier
     * authorization or a failure frame completes the flow with
     * [LinkClientException.Refused].
     * Cancelling collection cancels the call and waits for the blocking owner
     * to close its response source before collection finishes.
     */
    fun stream(endpoint: String, payload: Map<String, WireValue> = emptyMap()): Flow<WireValue> = flow {
        val body = Json.encodeToString(
            JsonElement.serializer(),
            kotlinx.serialization.json.buildJsonObject {
                put("args", kotlinx.serialization.json.buildJsonObject {
                    payload.forEach { (key, value) -> put(key, value.toJsonElement()) }
                })
            },
        ).toByteArray(Charsets.UTF_8)
        val path = "/link/stream/$endpoint"
        val frames = Channel<WireValue>(Channel.RENDEZVOUS)
        val retirement = LinkRetiredCancellation()
        val tracked = trackedCall(streamClient, request(path, body, signed = true)) { call ->
            frames.cancel(retirement)
            call.cancel()
        }
        val call = tracked.call
        var blockingOwner: Job? = null

        suspend fun settle(owner: Job?) {
            call.cancel()
            withContext(NonCancellable) {
                owner?.cancelAndJoin()
            }
        }

        try {
            supervisorScope {
                val owner = async(Dispatchers.IO) {
                    try {
                        call.execute().use { response ->
                            if (!response.isSuccessful) {
                                checkStatus(response.code, response.body.bytes())
                            }
                            response.body.source().use { source ->
                                while (true) {
                                    val line = source.readUtf8Line() ?: break
                                    if (line.isBlank()) continue
                                    val frame = DecodedLinkStreamFrame.fromJsonElement(
                                        Json.parseToJsonElement(line),
                                    )
                                    if (frame.isFailure) {
                                        throw LinkClientException.Refused(
                                            frame.code ?: "internal",
                                            frame.message ?: "stream failed",
                                        )
                                    }
                                    frames.send(frame.value ?: WireValue.NullValue)
                                }
                            }
                        }
                    } catch (failure: CancellationException) {
                        frames.cancel(failure)
                        throw failure
                    } catch (failure: LinkClientException) {
                        frames.close(failure)
                        throw failure
                    } catch (failure: Exception) {
                        currentCoroutineContext().ensureActive()
                        val carrier = LinkClientException.Carrier(0, failure.message ?: failure.javaClass.simpleName)
                        frames.close(carrier)
                        throw carrier
                    } finally {
                        frames.close()
                    }
                }
                blockingOwner = owner
                try {
                    try {
                        for (frame in frames) emit(frame)
                        owner.await()
                    } catch (_: LinkRetiredCancellation) {
                        throw LinkClientException.Carrier(0, "Link client is closed")
                    }
                } finally {
                    settle(owner)
                }
            }
        } finally {
            try {
                if (blockingOwner == null) settle(null)
            } finally {
                frames.cancel()
                finishTrackedCall(tracked)
            }
        }
    }

    /** Forget the paired identity; the host refuses the next request. */
    fun unpair() = store.clear()

    /** Request cancellation of every owned call and retire this client's
     * dispatcher and connections. Use [closeAndAwait] when subsequent work
     * must observe completed callbacks and stream collectors. */
    override fun close() {
        requestClose()
    }

    /** Retire the transport and wait for callbacks, stream collectors, and
     * the shared OkHttp dispatcher to reach quiescence. */
    suspend fun closeAndAwait() {
        val calls = requestClose()
        withContext(NonCancellable) {
            calls.forEach { tracked -> tracked.settled.await() }
            withContext(Dispatchers.IO) {
                transportClient.dispatcher.executorService.awaitTermination(Long.MAX_VALUE, TimeUnit.NANOSECONDS)
            }
        }
    }

    companion object {
        private val JSON_MEDIA_TYPE = "application/json".toMediaType()

        internal fun observed(
            baseUrl: String,
            pinnedFingerprint: String,
            store: LinkCredentialsStoring,
            transportConfig: LinkTransportConfig,
            callObserver: LinkCallObserver,
        ): LinkClient = LinkClient(baseUrl, pinnedFingerprint, store, transportConfig, callObserver)

        /** Rebuild the paired client from persisted credentials — the
         * relaunch path that skips pairing and pins the stored fingerprint
         * again. Null before the first successful pairing.
         * @param store persisted Link identity owner.
         * @param transportConfig deployment-owned HTTP timeouts.
         * @return the restored client, or null when no identity exists.
         */
        fun restore(store: LinkCredentialsStoring, transportConfig: LinkTransportConfig): LinkClient? {
            val credentials = store.load() ?: return null
            return LinkClient(
                baseUrl = credentials.endpoint,
                pinnedFingerprint = credentials.pinnedFingerprint,
                store = store,
                transportConfig = transportConfig,
            )
        }
    }

    private suspend fun post(path: String, body: ByteArray, signed: Boolean): ByteArray {
        val tracked = trackedCall(unaryClient, request(path, body, signed))
        try {
            return awaitResponse(tracked)
        } catch (failure: CancellationException) {
            throw failure
        } catch (failure: LinkClientException) {
            throw failure
        } catch (failure: Exception) {
            currentCoroutineContext().ensureActive()
            throw LinkClientException.Carrier(0, failure.message ?: failure.javaClass.simpleName)
        }
    }

    /** Enqueue a registered call; cancellation owns it before enqueue so a
     * concurrent [close] settles through callback failure or enqueue rejection. */
    private suspend fun awaitResponse(tracked: TrackedCall): ByteArray = suspendCancellableCoroutine { continuation ->
        val call = tracked.call
        val completed = AtomicBoolean(false)

        fun complete(result: Result<ByteArray>) {
            if (completed.compareAndSet(false, true) && continuation.isActive) continuation.resumeWith(result)
        }

        continuation.invokeOnCancellation { call.cancel() }
        val callback = object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                try {
                    complete(Result.failure(e))
                } finally {
                    finishTrackedCall(tracked)
                }
            }

            override fun onResponse(call: Call, response: Response) {
                try {
                    val responseBody = response.use { value ->
                        val bytes = value.body.bytes()
                        checkStatus(value.code, bytes)
                        bytes
                    }
                    complete(Result.success(responseBody))
                } catch (failure: Exception) {
                    complete(Result.failure(failure))
                } finally {
                    finishTrackedCall(tracked)
                }
            }
        }
        try {
            call.enqueue(callback)
        } catch (failure: Exception) {
            try {
                complete(Result.failure(failure))
            } finally {
                finishTrackedCall(tracked)
            }
        }
    }

    private fun request(path: String, body: ByteArray, signed: Boolean): Request {
        val builder = Request.Builder()
            .url(base + path)
            .post(body.toRequestBody(JSON_MEDIA_TYPE))
        if (signed) applyCredentials(builder, path, body)
        return builder.build()
    }

    private fun trackedCall(
        client: OkHttpClient,
        request: Request,
        retireCall: (Call) -> Unit = { call -> call.cancel() },
    ): TrackedCall = synchronized(lifecycleLock) {
        if (closed) throw LinkClientException.Carrier(0, "Link client is closed")
        TrackedCall(client.newCall(request), retireCall).also { tracked ->
            activeCalls.add(tracked)
        }
    }

    private fun finishTrackedCall(tracked: TrackedCall) {
        synchronized(lifecycleLock) {
            activeCalls.remove(tracked)
            tracked.settled.complete(Unit)
        }
    }

    private fun requestClose(): List<TrackedCall> {
        val calls = synchronized(lifecycleLock) {
            closed = true
            activeCalls.toList()
        }
        calls.forEach { tracked -> tracked.retire() }
        if (transportRetired.compareAndSet(false, true)) {
            transportClient.connectionPool.evictAll()
            transportClient.dispatcher.executorService.shutdown()
        }
        return calls
    }

    private fun applyCredentials(builder: Request.Builder, path: String, body: ByteArray) {
        val credentials = store.load() ?: throw LinkClientException.Unpaired()
        val privateKeyRaw = credentials.signingKeyRaw
            ?: throw LinkClientException.BadWire("stored signing key is not base64")
        val timestamp = System.currentTimeMillis().toString()
        val input = LinkSigning.signingInput(
            timestamp = timestamp,
            method = "POST",
            path = path,
            bodySha256Hex = LinkSigning.sha256Hex(body),
        )
        builder.header(LinkSigning.deviceIdHeader, credentials.deviceId)
        builder.header(LinkSigning.timestampHeader, timestamp)
        builder.header(LinkSigning.signatureHeader, LinkSigning.sign(input, privateKeyRaw))
    }

    private fun checkStatus(status: Int, body: ByteArray?) {
        if (status in 200..299) return
        val document = body?.let { bytes ->
            runCatching {
                Json.parseToJsonElement(bytes.decodeToString()) as? JsonObject
            }.getOrNull()
        }
        val message = document?.optionalString("message")?.takeIf { it.isNotEmpty() }
            ?: document?.optionalString("reason")?.takeIf { it.isNotEmpty() }
            ?: "HTTP $status"
        if (status == 403 && document?.optionalString("error") == "forbidden") {
            throw LinkClientException.Refused("forbidden", message)
        }
        throw LinkClientException.Carrier(status, message)
    }

}

private fun JsonObject.optionalString(field: String): String? =
    (this[field] as? JsonPrimitive)?.takeIf { it.isString }?.content

private fun JsonObject.requiredString(field: String): String =
    optionalString(field)?.takeIf { it.isNotEmpty() }
        ?: throw LinkClientException.BadWire("missing or invalid string field $field")

private fun JsonObject.requiredNumber(field: String): Double =
    (this[field] as? JsonPrimitive)?.takeUnless { it.isString }?.doubleOrNull
        ?: throw LinkClientException.BadWire("missing or invalid number field $field")

private fun JsonObject.pairResponse(expectedProtocolVersion: Double): LinkPairResponse {
    val role = requiredString("role")
    val protocolVersion = requiredNumber("linkProtocolVersion")
    if (protocolVersion != expectedProtocolVersion) {
        throw LinkClientException.BadWire("pair response protocol version mismatch")
    }
    return LinkPairResponse(
        deviceId = requiredString("deviceId"),
        hostId = requiredString("hostId"),
        hostName = requiredString("hostName"),
        role = LinkDeviceRole.entries.singleOrNull { it.wire == role }
            ?: throw LinkClientException.BadWire("pair response carried an invalid role"),
        linkProtocolVersion = protocolVersion,
    )
}

private fun JsonObject.boolean(field: String): Boolean =
    this[field]?.jsonPrimitive?.booleanOrNull ?: false

private fun JsonObject.number(field: String): Double =
    this[field]?.jsonPrimitive?.content?.toDoubleOrNull() ?: 0.0

/** Map the host-description JSON onto the generated contract model. */
internal fun mapHostDescription(obj: JsonObject): LinkHostDescription {
    val capabilities = obj["capabilities"]?.jsonObject ?: JsonObject(emptyMap())
    val session = capabilities["session"]?.jsonObject ?: JsonObject(emptyMap())
    val workspace = capabilities["workspace"]?.jsonObject ?: JsonObject(emptyMap())
    val interaction = capabilities["interaction"]?.jsonObject ?: JsonObject(emptyMap())
    return LinkHostDescription(
        linkProtocolVersion = obj.number("linkProtocolVersion"),
        contractVersion = obj.number("contractVersion"),
        hostVersion = obj.requiredString("hostVersion"),
        hostId = obj.requiredString("hostId"),
        hostName = obj.requiredString("hostName"),
        runtimeClass = obj.requiredString("runtimeClass"),
        sessionFormatVersion = obj.number("sessionFormatVersion"),
        allowRemoteApproval = obj.boolean("allowRemoteApproval"),
        capabilities = LinkCapabilities(
            session = LinkSessionCapabilities(
                list = session.boolean("list"),
                history = session.boolean("history"),
                follow = session.boolean("follow"),
                prompt = session.boolean("prompt"),
                cancel = session.boolean("cancel"),
            ),
            workspace = LinkWorkspaceCapabilities(follow = workspace.boolean("follow")),
            interaction = LinkInteractionCapabilities(
                approval = interaction.boolean("approval"),
                question = interaction.boolean("question"),
            ),
        ),
    )
}
